include/uuid/UUIDv8.h

```diff
@@ -0,0 +1,320 @@
+        for (int i = 0; i < 16; i++) {
+            auto e = (std::byte)uuid[i];
+            auto shifted_e = (std::byte)uuid[i];
+            shifted_e >>= 4;
+            str += DIGITS[(std::uint8_t)(shifted_e)];
+            str += DIGITS[(std::uint8_t)(e & (std::byte)0xf)];
+            if (i == 3 || i == 5 || i == 7 || i == 9) {
+                str += '-';
+            }
+        }
+        return str;
+    }
+
+    /**
+     * @brief  the given UUIDv8 string to UUIDv8 to 16 byte unsigned int array
+     * @param uuidStr - UUIDv8 equivalent in string format
+     * @return UUIDv8 object that contain 2 parts MSB and LSB
+     *
+     */
+    UUIDv8 fromString(std::string uuidStr);
+
+    /**
+     * @brief extracts UTC time at from current UUIDv8 object
+     * @return UTC time
+     */
+    uint64_t getTime() { return this->msb_ >> 16; }
+
+    /**
+     * @brief Utility function extracts UTC time from the given UUIDv8
+     * @param UUIDv8 object
+     * @return UTC time
+     */
+    static uint64_t getTime(UUIDv8 const& uuid) {
+        return uuid.msb_ >> 16;
+    }
+
+    /**
+     * @brief the random number that's part of UUIDv8 number
+     * @param UUIDv8
+     * @return  random number
+     */
+    uint64_t getRandom(const UUIDv8 uuid) { return uuid.lsb_ & randomMask_; }
+
+    /**
+     * @brief Returns version of UUIDv8 format
+     * @param UUIDv8 given UUIDv8 number in UUIDv8 object
+     * @return version
+     * @note The 4 bit UUIDv8 version (1000). Occupies bits 48 through 51.
+     */
+    uint64_t getVersion(UUIDv8 const& uuid) {
+        return ( uuid.msb_ >> 12) & 0xf;
+    }
+    /**
+     * @brief Returns 2-bit UUID variant of UUIDv8 format (10)
+     * @param UUIDv8 object
+     * @return variant of given UUIDv8 number format
+     */
+    uint64_t getVariant(UUIDv8 const& uuid) {
+        return (uuid.lsb_ >> 62) & 0x3;
+    }
+
+    /** @brief return current count of UUIDv8 numbers generated
+     * @return count
+    */
+    uint64_t getCount() { return (this->msb_ & 0xFFFL); }
+
+    /** @brief return count of UUIDv8 numbers generated from given UUIDv8
+     * @param UUIDv8 node
+     * @return count
+    */
+    uint64_t getCount(UUIDv8 const& uuid) { return (uuid.msb_ & 0xFFFL); }
+
+    /** @brief get MSB part from given UUIDv8
+     * @return msb_
+    */
+    uint64_t getMSB() { return msb_; }
+
+    /** @brief get LSB part from given UUIDv8
+     * @return lsb_
+      */
+    uint64_t getLSB() { return lsb_; }
+
+
+private:
+    /** uuidV8FromString - utility function to convert UUIDv8 string to uint8_t pointer array
+     * @param str input UUIDv8 string that needs to be converted to UUIdv8
+     * @param[out] uuidOut - will contain UUIDv8 number
+    */
+    inline int uuidV8FromString(std::string str, uint8_t* uuidOut) {
+        str.erase(remove(str.begin(), str.end(), '-'), str.end());
+        auto i = 0;
+        for (auto c : str) {
+            uint8_t n;
+            switch (c) {
```

Comment:
Can't we use library functions to convert ascii value to int

Reply:
The current logic is specific and will protect against any to any conversion of values. This logic only mandates conversions based on switch choices. Typecasting is broad way of doing it: any value to int. I would think, keeping the current logic would be good.